fvm/transactionInvoker.go

```diff
@@ -49,31 +49,41 @@ func (i TransactionInvoker) Process(
 
 	var modifiedSets programsCache.ModifiedSetsInvalidator
 	defer func() {
-		// based on the contract and frozen account updates we decide how to
-		// clean up the programs for failed transactions we also do the same as
-		// transaction without any deployed contracts
-		programs.Cleanup(modifiedSets)
-
 		if sth.NumNestedTransactions() > 1 {
-			err := sth.RestartNestedTransaction(nestedTxnId)
-			if err != nil {
+			if processErr == nil {
+				// This is a fvm internal programming error.  We should halt.
 				processErr = fmt.Errorf(
-					"cannot restart nested transaction: %w",
-					err,
-				)
+					"successfully executed transaction has unexpected " +
+						"nested transactions.")
+				return
+			} else {
+				err := sth.RestartNestedTransaction(nestedTxnId)
+				if err != nil {
+					// This should never happen since merging views should
+					// never fail.
+					processErr = fmt.Errorf(
+						"cannot restart nested transaction on error: %w "+
+							"(original txError: %v)",
+						err,
+						processErr)
+					return
+				}
+
+				ctx.Logger.Warn().
+					Msg("transaction has unexpected nested transactions")
+
+				// Note: proc.Err is set by TransactionProcedure.
+				proc.Logs = make([]string, 0)
+				proc.Events = make([]flow.Event, 0)
+				proc.ServiceEvents = make([]flow.Event, 0)
 			}
-
-			msg := "transaction has unexpected nested transactions"
-			ctx.Logger.Error().
-				Msg(msg)
-
-			proc.Err = errors.NewFVMInternalErrorf(msg)
```

Comment:
I'm going to remove this error/error code in a different PR.  This is currently classified as an error rather than a failure.  This should not be the case.